### ISSUE TITLE
issues when pm=ondemand

### ISSUE THREAD

**Issue by pbousek** [rvalitov/zabbix-php-fpm#11]
Hello,
I have several issues when process manager (pm) is set to ondemand
It's probably tied to way how ondemand work.

It seems that pools are found and lost again by discovery based on the fact that there is some process running.

And same situation cause trigger PHP-FPM uses queue on {HOST.NAME} pool {#POOLNAME} is flapping because the value differs (probably from none to some value).

Petr

**Comment by rvalitov** [comment]
Thank you, Petr. I think you are right, when we have ondemand then the script thinks that the pools are found and lost. I can think of an idea to make a cache of the pools, but in this case it's hard to say:
1) how long we should cache it? may be add a global configuration MACRO in Zabbix Server for that?
2) how to easily distinguish between ondemand pools and other pools? I suppose the only solution is to check the status of the pool in the discovery script, i.e. run the `zabbix_php_fpm_status.sh` and parse the pool type value.

Or simply update the trigger and make it related to `static` and `dynamic` pools only.

Any other ideas?

**Comment by pbousek** [comment]
Yes, zabbix_php_fpm_status.sh shows "process manager":"ondemand", but I dont know if cache is good solution. How long should be the cache active? In some cases when there is no traffic on the web, the cache must be active for really long time. 
Or it could be in cache until next value is received and if the pool socket is active.... don't know whats better.

**Comment by rvalitov** [comment]
I suppose the only stable method is to manually scan the PHP-FPM pool configuration files. But that's not a pleasant thing I would like to do 😅 

Another approach is to somehow detect if the pool belongs to `ondemand` and if that's true, then simply ignore this pool. I suppose that if a pool uses `ondemand` process manager, then this website does not have much traffic and may be offline most of the time. As a result, there's no sense in monitoring such pools in Zabbix. My idea of PHP monitoring is to detect if a pool goes offline, or if some pool consumes too much resources. I think that both these cases are not applicable for `ondemand` pools.

**Comment by pbousek** [comment]
Only thing that is disturbing on monitoring php-fpm with ondemand is that it often triggers "PHP-FPM uses queue on" when value for one check get value 1 and after that is for some time again 0
So maybe if you can somehow set to enable trigger "PHP-FPM uses queue on" only for pools that are not ondemand - is it possible?
Or ignore them at all... you are right that I use them for not important web services with small trafiic, because it allocates less memory.

**Comment by rvalitov** [comment]
@pbousek I made a new PR #14 , please download the new Zabbix template from there and test it. Waiting for your feedback there.

**Comment by pbousek** [comment]
@rvalitov it seems that its fine now.... ondemand are not triggered now... but what exactly means the "listen queue" value? Whats wrong with php when it has value 1? I still have alarms from php-fpm that some pool "PHP-FPM uses queue" but I didn't exacty find any problem with php, everything works fine, log files are empty. What should tell me this trigger? Maybe... if this value indicate number of requests beeing queued before processed, than it will be better to trigger it only when more than only single value differs from 0, because it usually starts new process, request is processed and next value is 0 again. Sorry for my english, hope you understand me.

**Comment by rvalitov** [comment]
@pbousek 
Thank you for your tests and comments. Let me explain the `listen queue` parameter and the `PHP-FPM uses queue` warning. If the queue value is greater than zero, then it means that there are pending connections that are not processed by PHP-FPM. It can be interpreted as:
1. If the values are low and `listen queue` is lower than `max queue length` (I suppose this is your case), then it's a performance issue. It means that your PHP-FPM does not process the connections for this pool fast enough, as a result it has to use the queue. If you have low values, like `1` or `3` here, then it's not a severe problem, because your queue is not full, but it's not empty. As a result, the website works fine (no errors in PHP log, Apache or nginx log), no visual glitches on website, but the performance is degraded - for example, your website could load in 1 second, but because of the queue the loading time becomes longer, 2 seconds or more. It's like a swap file on HDD - if you are low on RAM memory, then swap file is used, but the more swap space you use, the slower your server becomes. The same is with this `queue` parameter. The more queue is used, the slower the PHP/wesbite works.
2. If the queue becomes full, i.e. the results reach the `max queue length` parameter, then it's severe error. In this case the queue and PHP can't accept new connections and the new connections will be rejected. As a result the Apache or nginx will think that the PHP is not available (no response situation) and will log them in the error log file. If a user tries to access your website while the queue is full, then he will see a "Server not available" or a similar HTTP 500 error.

I think in your case the queue value may be greater than 1 if many users try to access your website. You can use [siege utility](https://packages.debian.org/buster/siege) to make a benchmark test.

**Comment by rvalitov** [comment]
> it will be better to trigger it only when more than only single value differs from 0, because it usually starts new process, request is processed and next value is 0 again
Yes, that's a good idea. The trigger should appear only when the queue value becomes greater than 0. The trigger should not appear when the value returns back to zero.

**Comment by rvalitov** [comment]
More info about the queue trigger:

> The current number of connections that have been initiated, but not yet accepted. If this value is non-zero it typically means that all the available server processes are currently busy, and there are no processes available to serve the next request. Raising pm.max_children (provided the server can handle it) should help keep this number low. This property follows from the fact that PHP-FPM listens via a socket (TCP or file based), and thus inherits some of the characteristics of sockets.

**Comment by pbousek** [comment]
@rvalitov thank you, so I understand it good, but all my triggers are situation when listen queue is "0" all the time and than one check returns "1" and the next one "0" again, so I think that it can occur in situation when all started processes are busy and it waits to raise new one?? 
But in real life, I would like to be informed when the listen queue is greater than 0 more than one check, lets say it could raise an alarm when last 3 checks are greater than zero? What do you think?

**Comment by pbousek** [comment]
And I don't know - issue #17 - I think that it works as expected, trigger is sent only when (in my case) listen queue change value 0 -> 1 and is resolved when value returns back to 0 (in my case next one check, so last 3 values are 0, 1, 0 )

**Comment by rvalitov** [comment]
@pbousek thank you for the info! Yes, probably triggering only after the last 3 checks failed is a good approach. I will go this way. 

Regarding the #17, I checked my code in the template, and the bug is there. The trigger is sensitive to the changes of the queue value, and should modified to check for any positive values instead. In your case the alarm is raised only once, because:

- the settings of the trigger, the alarm is raised only once to avoid too many messages, even if multiple trigger events happen
- the queue value returns back to zero quickly (after a couple of checks)

I will update the code and let you know when it's ready for next test.

**Comment by pbousek** [comment]
@rvalitov maybe there is another bug, i found that triggers are somehow multiplicated, but don't know why
![image](https://user-images.githubusercontent.com/21953786/71727006-3c6cd380-2e39-11ea-851f-c1d2d79e56be.png)
it seems that localhost:port value has changed for discovery rule, so new item/triggers are added

**Comment by pbousek** [comment]
it seems, that it somehow thinks that port changed from 8001 to another so it detect new pool, but after that, pool is not active anymore
![image](https://user-images.githubusercontent.com/21953786/71727479-a8037080-2e3a-11ea-8510-35a54f032b2e.png)

**Comment by pbousek** [comment]
here is the problem
```
root@api1:/etc/zabbix# lsof -p 26838 |grep -e unix -e TCP 
php-fpm7. 26838 www-data    9u  IPv4 224352639      0t0       TCP localhost:8001 (LISTEN)
root@api1:/etc/zabbix# lsof -p 26838 |grep -e unix -e TCP 
php-fpm7. 26838 www-data    9u  IPv4 224352639      0t0       TCP localhost:8001 (LISTEN)
root@api1:/etc/zabbix# lsof -p 26838 |grep -e unix -e TCP 
php-fpm7. 26838 www-data    9u  IPv4 224352639      0t0       TCP localhost:8001 (LISTEN)
root@api1:/etc/zabbix# lsof -p 26838 |grep -e unix -e TCP 
php-fpm7. 26838 www-data    4u  IPv4 324700387      0t0       TCP localhost:8001->localhost:51786 (ESTABLISHED)
php-fpm7. 26838 www-data    9u  IPv4 224352639      0t0       TCP localhost:8001 (LISTEN)
```

you get the first value - that only say that php process has established connection from 8001 -> 51786

but you should work with the LISTEN line

**Comment by rvalitov** [comment]
@pbousek thank you for this info. I will check that. Please, make a separate issue about the last (new) problem. One issue per problem please. Thank you!

**Comment by pbousek** [comment]
sorry for that, I though at first that the problem is somehow related to ondemand, but its problem of parsing lsof output, new issue opened
I've deleted it

**Comment by rvalitov** [comment]
@pbousek I updated the code, now the trigger alarms only if the queue is greater than zero during the last 3 checks. I closed this issue, but your feedback is most welcome. Meanwhile I proceed to your next issue reported.

**Comment by pbousek** [comment]
OK, I import the new template and let you know if it works as desired
But how to solve detection of ondemand pools that then disappear from discovered pools - it's listed in items with "The item is not discovered anymore and will be deleted in 29d 23h 56m (on 2020-02-03 at 00:07)."
I think that "ondemand" problem is not solved finally.

**Comment by rvalitov** [comment]
I have an idea here. If the discovery script detects the `ondemand` pool, then it can cache it and check its status. If the pool responds to a status call, then the pool is still active (enabled). Otherwise the pool is disabled or deleted. I checked such approach for `ondemand` pools that listen to TCP and socket. However, in this case if the pool does not have any requests and is "sleeping", then triggering it with the status check will start the pool process (fork children). As a result, the `ondemand` pool will:

- either be always "active" (never sleeps), i.e. will have at least 1 process
- or will repeatedly go into "active" and "sleep" phases, i.e. when it starts 1 process during the status check initiated by Zabbix agent, and then stops all processes when the Zabbix agent sleeps until the next check.

The behavior above depends on the value of `pm.process_idle_timeout` setting.

The discovery script can detect only those pools that have any active children or processes. So, sleeping pools can not be detected (discovered) until they become active, i.e. have some processes and traffic.

What do you think about that?

**Comment by pbousek** [comment]
template doesn't work as expected

https://www.zabbix.com/documentation/4.4/manual/appendix/triggers/functions
last(#3) - third most recent value (not three latest values)

so you just move trigger to third value in history, you not check last three values

**Comment by pbousek** [comment]
I think that you must use
{Template App PHP-FPM:php-fpm.listen_queue[{#POOLSOCKET}].last()}>0 and {Template App PHP-FPM:php-fpm.listen_queue[{#POOLSOCKET}].last(#2)}>0 and {Template App PHP-FPM:php-fpm.listen_queue[{#POOLSOCKET}].last(#3)}>0
it validates last three values

**Comment by rvalitov** [comment]
Yes, my mistake. Thank you for pointing it.
There's an item `php-fpm.queue` - queue utilization. I think it can help us here.
So, I have an idea to remove the trigger "pool uses queue" which is not informative enough and instead add the following:

|Name|Severity|Description|
|------|------|------|
|The queue is not available|Warning|The queue is not available (max queue length is zero), i.e. it was not specified by the user in the pool's configuration (`backlog` option). As a result, some requests may be dropped or rejected. It's recommended to configure a queue for each pool.|
|The queue utilization for pool reached 25%|Warning|The queue for this pool reached 25% of its maximum capacity. Items in queue represent the current number of connections that have been initiated on this pool, but not yet accepted. It typically means that all the available server processes are currently busy, and there are no processes available to serve the next request. Raising pm.max_children (provided the server can handle it) should help keep this number low. This trigger follows from the fact that PHP-FPM listens via a socket (TCP or file based), and thus inherits some of the characteristics of sockets. Low values of the listen queue generally result in performance issues of this pool.|
|The queue utilization for pool reached 50%|Average|The queue for this pool reached 50% of its maximum capacity. Items in queue represent the current number of connections that have been initiated on this pool, but not yet accepted. It typically means that all the available server processes are currently busy, and there are no processes available to serve the next request. Raising pm.max_children (provided the server can handle it) should help keep this number low. This trigger follows from the fact that PHP-FPM listens via a socket (TCP or file based), and thus inherits some of the characteristics of sockets. Low values of the listen queue generally result in performance issues of this pool. Please, check that your server has enough resources and adjust the configuration of this pool to handle more concurrent requests, otherwise you can suffer serious degraded performance.|
|The queue utilization for pool reached 85%|High|The queue for this pool reached 85% of its maximum capacity. Items in queue represent the current number of connections that have been initiated on this pool, but not yet accepted. It typically means that all the available server processes are currently busy, and there are no processes available to serve the next request. Raising pm.max_children (provided the server can handle it) should help keep this number low. This trigger follows from the fact that PHP-FPM listens via a socket (TCP or file based), and thus inherits some of the characteristics of sockets. Low values of the listen queue generally result in performance issues of this pool. Please, check that your server has enough resources and adjust the configuration of this pool to handle more concurrent requests, otherwise you can face severe errors when new requests can't be processed and will be rejected generating errors such as HTTP 500.|

I think such approach is more robust and provides more information about what's going on. What do you think about that?

**Comment by rvalitov** [comment]
@pbousek please test this PR #19. Please, keep in mind that many files were modified here, and should be replaced.

**Comment by pbousek** [comment]
@rvalitov PR #19 tested and seems good
will you implement you last idea about queue? Sounds good to me.

**Comment by rvalitov** [comment]
@pbousek thank you very much for testing! I will implement the new triggers soon. I will inform you when ready.

**Comment by rvalitov** [comment]
@pbousek please test a new PR #20 
Thank you!

**Comment by pbousek** [comment]
@rvalitov seems to be fine, I don't know how to simulate the issue to check the trigger but at least it doesn't trigger any fail positive queue issues

**Comment by rvalitov** [comment]
@pbousek thank you so much for all your help and time! I'm glad we could solve many issues together and improve the template. Now it seems we reached a solid "stable" version of the template. If you will have any problems or new ideas, please, don't hesitate to open new issues.
